lib/rrdd_plugin.ml

```diff
@@ -11,399 +11,396 @@
+    D.debug "Forking command %s" cmdstring;
+    (* create pipe for reading from the command's output *)
+    let (out_readme, out_writeme) = Unix.pipe () in
+    let cmd, args = match Astring.String.cuts ~empty:false ~sep:" " cmdstring with [] -> assert false | h::t -> h,t in
+    let pid = Forkhelpers.safe_close_and_exec None (Some out_writeme) None [] cmd args in
+    Unix.close out_writeme;
+    let in_channel = Unix.in_channel_of_descr out_readme in
+    let vals = ref [] in
+    let rec loop () =
+      let line = input_line in_channel in
+      let ret = f line in
+      begin
+        match ret with
+        | None -> ()
+        | Some v -> vals := v :: !vals
+      end;
+      loop ()
+    in
+    (try loop () with End_of_file -> ());
+    Unix.close out_readme;
+    let (pid, status) = Forkhelpers.waitpid pid in
+    begin
+      match status with
+      | Unix.WEXITED n   -> D.debug "Process %d exited normally with code %d" pid n
+      | Unix.WSIGNALED s -> D.debug "Process %d was killed by signal %d" pid s
+      | Unix.WSTOPPED s  -> D.debug "Process %d was stopped by signal %d" pid s
+    end;
+    List.rev !vals
 end
 
 (* Establish a XMLPRC interface with RRDD *)
 module RRDD = Rrd_client.Client
 
 module Xs = struct
-	module Xs = Xs_client_unix.Client(Xs_transport_unix_client)
-	include Xs
-
-	type xs_state = {
-		my_domid: int32;
-		root_path: string;
-		client: Xs.client;
-	}
-
-	let cached_xs_state = ref None
-
-	let cached_xs_state_m = Mutex.create ()
-
-	let get_xs_state () =
-		Mutex.execute cached_xs_state_m
-			(fun () ->
-				match !cached_xs_state with
-				| Some state -> state
-				| None ->
-					(* This creates a background thread, so must be done after daemonising. *)
-					let client = Xs.make () in
-					let my_domid =
-						Xs.immediate
-							client
-							(fun handle -> Xs.read handle "domid")
-						|> Int32.of_string
-					in
-					let root_path = Printf.sprintf "/local/domain/%ld/rrd" my_domid in
-					let state = {
-						my_domid;
-						root_path;
-						client
-					}
-					in cached_xs_state := Some state;
-					state)
+  module Xs = Xs_client_unix.Client(Xs_transport_unix_client)
+  include Xs
+
+  type xs_state = {
+    my_domid: int32;
+    root_path: string;
+    client: Xs.client;
+  }
+
+  let cached_xs_state = ref None
+
+  let cached_xs_state_m = Mutex.create ()
+
+  let get_xs_state () =
+    Mutex.execute cached_xs_state_m
+      (fun () ->
+         match !cached_xs_state with
+         | Some state -> state
+         | None ->
+           (* This creates a background thread, so must be done after daemonising. *)
+           let client = Xs.make () in
+           let my_domid =
+             Xs.immediate
+               client
+               (fun handle -> Xs.read handle "domid")
+             |> Int32.of_string
+           in
+           let root_path = Printf.sprintf "/local/doxen-api/main/%ld/rrd" my_domid in
```

Comment:
`domain -> doxen-api/main` Is this deliberate?

Reply:
No, well spotted